specs/marketplace.md

```diff
@@ -0,0 +1,381 @@
+consider the node's operator configuration when making the decision. The storage provider node that originally hosted
+the freed slot MAY also participate in the data repair, but my refilling the slot it **won't** recover its original collateral
+and needs to submit new collateral with the `fillSlot()` call.
+
+The repair process is the same as with the filling slots, with one difference that the node MUST use the erasure coding to 
+reconstruct the original dataset. As this requires retrieving more data of the dataset from the network, the node that 
+will successfully repair slot by filling the freed slot will be granted additional reward. (**TODO: Implementation**)
+
+The repair process is then as follows:
+
+1. Node detects `SlotFreed` event and decide to repair it.
+1. Node MUST download the required chunks and MUST use the [Reed-Soloman algorithm](https://hackmd.io/FB58eZQoTNm-dnhu0Y1XnA) to reconstruct the original slot's data.
+1. Node MUST generate proof over the reconstructed data.
+1. Node MUST submit transaction with call to `fillSlot()` with the same parameters and collateral allowance as described in the [Filling slot](#filling-slot).
+
+### Collecting funds
+
+Storage Provider node SHOULD monitor Requests and slots it hosts. In case it needs to discover what slots it is hosting,
+for example, because the node had to restart, then it SHOULD use the contract call `mySlots()`, which returns array of `SlotID`s
+associated with the Ethereum address from which the contract call originates. This list is kept up to date by the smart contract itself.
+
+When node slot's Requests reaches states `Cancelled`, `Finished` or `Failed` it SHOULD call the contract's `freeSlot(slotId)` function.
+These states can be detected using:
+
+ - `Cancelled` state MAY be detected by setting timeout using `expiry` **and** not detecting `RequestFulfilled(requestId)` event. There is also `RequestCancelled` event emitted, but the node SHOULD NOT use it for asserting expiry as it is not guaranteed to be emitted at the time of expiry.
+ - `Finished` state MAY be detected by setting timeout specified from function `getRequestEnd(requestId)`.
+ - `Failed` state MAY be detected by listening to the `RequestFailed(requestId)` event emitted.
+
+For each of these states, different funds are collected:
+
+- For `Cancelled` the collateral is returned together with proportional payout based on time that the node actually hosted the dataset before expiry was reached.
+- For `Finished` the full reward for hosting the slot together with collateral is gathered.
+- For `Failed` no funds are collected as reward is returned to the client and collateral is burned, but this call removes the slot from the `mySlots()` tracking.
+
+## Validator role
+
+Validator role represents nodes that verify that the Storage Provider nodes submit proofs when they are required.
```

Comment:
Perhaps it's good to explicitly mention that the validator doesn't validate anything. It's the contract that does the actual validation that a proof is missing, it only needs the validator to invoke the contract in a timely manner.